rules/validator/namespace/kv/validator.go

```diff
@@ -0,0 +1,152 @@
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package kv
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/m3db/m3cluster/kv"
+	kvutil "github.com/m3db/m3cluster/kv/util"
+	"github.com/m3db/m3metrics/rules/validator/namespace"
+	"github.com/m3db/m3x/log"
+)
+
+var (
+	errValidatorClosed = errors.New("validator is closed")
+)
+
+type validator struct {
+	sync.RWMutex
+
+	opts               NamespaceValidatorOptions
+	logger             log.Logger
+	kvStore            kv.Store
+	validNamespacesKey string
+	initWatchTimeout   time.Duration
+
+	closed          bool
+	doneCh          chan struct{}
+	wg              sync.WaitGroup
+	validNamespaces map[string]struct{}
+}
+
+// NewNamespaceValidator creates a new namespace validator.
+func NewNamespaceValidator(opts NamespaceValidatorOptions) (namespace.Validator, error) {
+	v := &validator{
+		opts:               opts,
+		logger:             opts.InstrumentOptions().Logger(),
+		kvStore:            opts.KVStore(),
+		validNamespacesKey: opts.ValidNamespacesKey(),
+		initWatchTimeout:   opts.InitWatchTimeout(),
+		doneCh:             make(chan struct{}),
+		validNamespaces:    toStringSet(opts.DefaultValidNamespaces()),
+	}
+	if err := v.watchRuntimeConfig(); err != nil {
+		return nil, err
+	}
+	return v, nil
+}
+
+// Validate validates whether a given namespace is valid.
+func (v *validator) Validate(ns string) error {
+	v.RLock()
+	defer v.RUnlock()
+
+	if v.closed {
+		return errValidatorClosed
+	}
+	if _, exists := v.validNamespaces[ns]; !exists {
+		return fmt.Errorf("%s is not a valid namespace", ns)
+	}
+	return nil
+}
+
+func (v *validator) Close() {
+	v.Lock()
+	if v.closed {
+		v.Unlock()
+		return
+	}
+	v.closed = true
+	close(v.doneCh)
+	v.Unlock()
+
+	// NB: Must wait outside the lock to avoid a circular dependency
+	// between the goroutine closing the validator and the goroutine
+	// processing updates.
+	v.wg.Wait()
+}
+
+func (v *validator) watchRuntimeConfig() error {
+	watch, err := v.kvStore.Watch(v.validNamespacesKey)
+	if err != nil {
+		return err
+	}
+	kvOpts := kvutil.NewOptions().SetLogger(v.logger)
+	select {
+	case <-watch.C():
+		v.processUpdate(watch.Get(), kvOpts)
+	case <-time.After(v.initWatchTimeout):
+		v.logger.WithFields(
+			log.NewField("key", v.validNamespacesKey),
+			log.NewField("timeout", v.initWatchTimeout),
+			log.NewField("default", v.opts.DefaultValidNamespaces()),
+		).Warnf("timed out waiting for initial valid namespaces")
+	}
+
+	v.wg.Add(1)
+	go func() {
+		defer v.wg.Done()
+
+		for {
+			select {
+			case <-watch.C():
+				v.processUpdate(watch.Get(), kvOpts)
+			case <-v.doneCh:
+				return
+			}
+		}
+	}()
+	return nil
+}
+
+func (v *validator) processUpdate(value kv.Value, opts kvutil.Options) {
+	strs, err := kvutil.StringArrayFromValue(value, v.validNamespacesKey, v.opts.DefaultValidNamespaces(), opts)
+	if err != nil {
+		// kvutil already logged the error.
```

Comment:
This is unrelated to this diff, but we should we remove logging from `kvutil`? If we're returning the error, wouldn't the caller be the best place to log any errors?

Reply:
Possibly although not my change, agreed that it's probably cleaner to log on the client side, but to remove it now it would require finding all such places and adding logs accordingly.

Reply:
Yea, didn't mean to imply we should do it now. Just figured I would bring it up for discussion.

Reply:
Yeah I think client can still log the error if needed to provide more context around the error, since the logging done inside kvutil is at a lower level.